src/dune/cached_digest.ml

```diff
@@ -26,6 +26,11 @@ end)
 
 let needs_dumping = ref false
 
+(* CR-soon amokhov: replace this mutable table with a memoized function. This
+   will probably require splitting this module in two, for dealing with source
+   and target files, respectively. For source files, we receive updates via the
+   file-watching API. For target files, we modify the digests ourselves, without
+   subscribing for file-watching updates. *)
```

Comment:
One thing to be mindful about here: memoized functions are currently not persistent, while cached digests for targets are. If we used a memoised function for target digests, I'd expect null builds to get slower.